test/adapter/json_api/errors_test.rb

```diff
@@ -0,0 +1,154 @@
+require 'test_helper'
+
+module ActiveModel
+  class Serializer
+    class Adapter
+      class JsonApi
+        class ErrorsTest < Minitest::Test
+
+          # see
+          # https://github.com/rails/rails/blob/4-2-stable/activemodel/lib/active_model/errors.rb
+          # The below allows you to do:
+          #
+          #   model = ModelWithErrors.new
+          #   model.validate!            # => ["cannot be nil"]
+          #   model.errors.full_messages # => ["name cannot be nil"]
+          class ModelWithErrors
+            # Required dependency for ActiveModel::Errors
+            extend ActiveModel::Naming
+            def initialize
+              @errors = ActiveModel::Errors.new(self)
+            end
+            attr_accessor :name
+            attr_reader   :errors
+
+            # The following methods are needed to be minimally implemented
+
+            def read_attribute_for_validation(attr)
+              send(attr)
+            end
+
+            def self.human_attribute_name(attr, options = {})
+              attr
+            end
+
+            def self.lookup_ancestors
+              [self]
+            end
+          end
+
+          def setup
+            @model = ModelWithErrors.new
+          end
+
+          class JsonApiErrorSerializer < ActiveModel::Serializer
+            attribute :source
+            attribute :detail
+
+            def source
+              {pointer: "data/attributes/#{object.first}"}
+            end
+
+            def detail
+              detail = object.last
+              if detail.respond_to?(:join)
+                detail.join
+              else
+                detail
+              end
+            end
+          end
+
+          class JsonApiErrorsSerializer < ActiveModel::Serializer
+            has_many :errors, serializer: JsonApiErrorSerializer
+
+            def errors
+              object.messages.to_a
+            end
+          end
+
+          class ActiveModel::Serializer::Adapter::JsonApiError < ActiveModel::Serializer::Adapter::FlattenJson
+          end
+
+          def test_active_model_errors
+            @model.errors.add(:name, "cannot be nil")
+            serializer_opts = {serializer: JsonApiErrorSerializer}
+            errors = @model.errors
+
+            serializer = JsonApiErrorSerializer.new(errors.first, serializer_opts)
+            expected_error_object =
+            {
+              source: { pointer: 'data/attributes/name' },
+              detail: 'cannot be nil'
+            }
```

Comment:
@bf4 Although it is still compliant to represent multiple validation errors in the same error object, it's not ideal. I would recommend creating a unique error object per validation error, so that fields such as `code` and `title` can be used consistently and expressively across errors. For instance, one `code` might indicate that a field is too short, and another might indicate that invalid characters were used. If you try to combine these errors into a single `code`, then it will probably end up being something more generic with less meaning, such as "invalid value".